src/file_tree.ml

```diff
@@ -70,48 +64,52 @@ module Dune_file = struct
     | Plain         x -> x.path
     | Ocaml_script  p -> p
 
-  let extract_subdirs ~project sexps =
-    let subdirs, ignored_subdirs, sexps =
-      List.fold_left sexps ~init:(None, [], [])
-        ~f:(fun (subdirs, ignored, sexps) sexp ->
-        match (sexp : Dune_lang.Ast.t) with
-        | List (loc, (Atom (_, A ("ignored_subdirs" | "subdirs")) :: _)) ->
-          let stanza =
-            Dune_project.set_parsing_context project Subdirs_stanza.decode in
-          let subdir = Dune_lang.Decoder.parse stanza Univ_map.empty sexp in
-          begin match subdir, subdirs, ignored with
-          | Ignore_subdirs i, None, is -> (subdirs, i :: is, sexps)
-          | Subdirs x, None, [] -> (Some x, [], sexps)
-          | Subdirs _, Some _, _ ->
-            Errors.fail loc "More than one subdirs stanza is not allowed"
-          | Ignore_subdirs _, Some _, _
-          | Subdirs _, _, _::_ ->
-            Errors.fail loc "Cannot have both subdirs and ignored_subdirs \
-                             stanza in a dune file. "
-          end
-        | _ -> subdirs, ignored, sexp :: sexps)
+  let extract_sub_dirs ~project sexps =
+    let (sub_dirs, ignored_sub_dirs, data_only), sexps =
+      List.fold_left sexps ~init:((None, [], None), [])
+        ~f:(fun ((sub_dirs, ignored, data_only) as stanzas, sexps) sexp ->
+          match (sexp : Dune_lang.Ast.t) with
+          | List ( loc , (Atom (_ , A ("ignored_subdirs"
+                                      | "subdirs" | "data_only")) :: _)) ->
+            let stanza =
+              Dune_project.set_parsing_context project Sub_dirs_stanza.decode in
+            let stanza = Dune_lang.Decoder.parse stanza Univ_map.empty sexp in
+            let stanzas =
+              match stanza, sub_dirs, data_only, ignored with
+              | Ignore_sub_dirs i, None, _, _ ->
+                (sub_dirs, i :: ignored, data_only)
+              | Data_only x, _, None, _ -> (sub_dirs, ignored, Some x)
+              | Sub_dirs x, None, _, [] -> (Some x, [], data_only)
+              | Sub_dirs _, Some _, _, _ ->
+                Errors.fail loc "More than one sub_dirs stanza is not allowed"
+              | Ignore_sub_dirs _, Some _, _, _
+              | Sub_dirs _, _, _, _::_ ->
+                Errors.fail loc
+                  "Cannot have both sub_dirs and ignored_sub_dirs \
+                   stanza in a dune file. "
+              | Data_only _, _, Some _, _ ->
+                Errors.fail loc "More than one data_only stanza is not allowed"
```

Comment:
A slightly nicer way to write this code would be something like this:

```ocaml
fields (
  let%map ignored_sub_dirs = multi_field "ignored_subdirs" ...
  and sub_dirs = field "sub_dirs" ...
  and data_only = field "data_only"
  and rest = leftover_fields
  in
 ...
)
```

where `leftover_field` is a new combinator to add that would return the unparsed fields as a `Sexp.Ast.t list`.

Reply:
I tried doing this but i can't get it to work. The last commit is WIP, will try again in a bit.